### PR TITLE
cilium, ipvlan: add iptables rules switch and set up ipvlan mode accordingly

### PR DESCRIPTION
**Pull request by borkmann**
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6662)

### CODE REVIEW

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

File: daemon/daemon_main.go [hunk 1]
```diff
@@ -715,6 +718,10 @@ func initEnv(cmd *cobra.Command) {
 		loadinfo.StartBackgroundLogger()
 	}
 
+	if !option.Config.InstallIptRules {
+		option.Config.Masquerade = false
```
Comment (brb) [inline daemon/daemon_main.go:722]: I think this is redundant, as `iptables.InstallRules()` installs the MASQ rules and it is not called in the case of `option.Config.InstallIptRules = false`.

File: pkg/endpoint/connector/ipvlan.go [hunk 2]
```diff
@@ -215,29 +215,32 @@ func getIpvlanMasterName() string {
...
-func SetupIpvlan(id string, mtu int, masterDev int, ep *models.EndpointChangeRequest) (*netlink.IPVlan, *netlink.Link, string, error) {
+func SetupIpvlan(id string, mtu, masterDev int, mode netlink.IPVlanMode, ep *models.EndpointChangeRequest) (*netlink.IPVlan, *netlink.Link, string, error) {
 	if id == "" {
 		return nil, nil, "", fmt.Errorf("invalid: empty ID")
 	}
 	if masterDev == 0 {
 		return nil, nil, "", fmt.Errorf("invalid: master device ifindex")
 	}
+	if mode != netlink.IPVLAN_MODE_L3 && mode != netlink.IPVLAN_MODE_L3S {
+		return nil, nil, "", fmt.Errorf("invalid: ipvlan operation mode")
```
Comment (brb) [inline pkg/endpoint/connector/ipvlan.go:226]: Maybe it would be good to report the invalid mode in the error msg?

**Review by brb — COMMENTED** [review]
LGTM besides a few nit-picks.

File: pkg/endpoint/connector/ipvlan.go (hunk 2 above)
Comment (borkmann) [reply pkg/endpoint/connector/ipvlan.go:226]: It's actually printed in cilium-cni.go which takes the opmode string and translates it to netlink.IPVLAN_MODE_* enum. Unsupported mode string will be dumped there, the one above is internal from CNI side only.

File: daemon/daemon_main.go (hunk 1 above)
Comment (borkmann) [reply daemon/daemon_main.go:722]: Yes correct, though wanted to make sure here that noone else is making an assumption about only option.Config.Masquerade in future, hence implicit update.

**Review by brb — APPROVED** [review]

**Review by raybejjani — APPROVED** [review]

**Pull request merged**